### PR TITLE
fix: add handling for struct outputs

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
Adds handling for methods that output struct values

fixes: #574
fixes: #573 
requires: https://github.com/ApeWorX/ethpm-types/pull/28

### How I did it
If we detect that a return value is a struct (a tuple with members that have names), then create a dataclass-based "struct definition" and return the data inside of that ad-hoc struct.

### How to verify it
Create a contract with structs in it:

```vyper
struct Test:
    a: address
    b: bytes32

struct Nested:
    t: Test

@view
@external
def foo() -> Test:
    return Test({a: msg.sender, b: block.prevhash})


@view
@external
def bar() -> Nested:
    return Nested({t: Test({a: msg.sender, b: block.prevhash})})
```

then try it:
```py
>>> r = c.foo()
>>> r.a  # use like a struct
'0x1e59ce931b4cfea3fe4b875411e280e173cb7a9c'
>>> r[0]  # use like a tuple too
'0x1e59ce931b4cfea3fe4b875411e280e173cb7a9c'
# NOTE: Output values need further processing to become real addresses
```

### Checklist

- [ ] Handle nested structs
- [ ] Consider allowing structs to have dict-like access too (e.g. `r["a"]` in example)
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape_ethereum/ecosystem.py [hunk 1]
```diff
@@ -154,25 +155,56 @@ def decode_returndata(self, abi: MethodABI, raw_data: bytes) -> Tuple[Any, ...]:
...
                     value = self.decode_address(value)
+                except InsufficientDataBytes as err:
+                    raise DecodingError() from err
+
+            output_values.append(value)
+
+        if (
+            len(abi.outputs) == 1
+            and abi.outputs[0].components
+            and all(c.name != "" for c in abi.outputs[0].components)
```
Comment (NotPeopling2day) [inline src/ape_ethereum/ecosystem.py:181]: Currently failing `error: "ABIType" has no attribute "components"`

Comment (fubuloubu) [reply src/ape_ethereum/ecosystem.py:181]: requires https://github.com/ApeWorX/ethpm-types/pull/28 to merge

File: src/ape_ethereum/ecosystem.py [hunk 2]
```diff
@@ -154,25 +155,56 @@ def decode_returndata(self, abi: MethodABI, raw_data: bytes) -> Tuple[Any, ...]:
...
+                        "__getitem__": lambda self, index: tuple(
+                            getattr(self, field) for field in self.__dataclass_fields__
+                        ).__getitem__(index)
+                    },
+                )
 
-                output_values.append(value)
+            struct_def = create_struct(
+                # NOTE: unnamed output structs appear as tuples with named members,
+                #       but we don't know of what struct type they are (because ABI encoding sucks)
```
Comment (unparalleled-js) [inline src/ape_ethereum/ecosystem.py:199]:
```suggestion
                #       but we don't know of what struct type they are (because of ABI encoding limitations)
```

File: src/ape_ethereum/ecosystem.py [hunk 3]
```diff
@@ -154,25 +155,56 @@ def decode_returndata(self, abi: MethodABI, raw_data: bytes) -> Tuple[Any, ...]:
...
+
+            output_values.append(value)
+
+        if (
+            len(abi.outputs) == 1
+            and abi.outputs[0].components
+            and all(c.name != "" for c in abi.outputs[0].components)
+        ):
+
+            def create_struct(name: str, types: List[ABIType]):
```
Comment (unparalleled-js) [inline src/ape_ethereum/ecosystem.py:184]: Since this is only used the once directly after, we might not need this as a method

**Review by unparalleled-js — COMMENTED** [review]
I can do some testing on this tomorrow!

File: src/ape_ethereum/ecosystem.py (hunk 3 above)
Comment (fubuloubu) [reply src/ape_ethereum/ecosystem.py:184]: This might be a good utility function, to have a standard way of describing ABI struct defs

**Comment by fubuloubu** [comment]
> I can do some testing on this tomorrow!

This PR is starting off as a proof-of-concept, feel free to make any changes necessary to clean up the code or improve it's functionality!

**Comment by unparalleled-js** [comment]
One edge case I've noticed is that:

```solidity
    function getSecret() public onlyOwner view returns(uint256 secret) {
        return 123;
    }
```

has the same output ABI as a struct with a single item in it.

It makes sense to just return the single value in this case, but the user loses that (pointless) name given to it

**Comment by unparalleled-js** [comment]
One edge case I've noticed is that:

```solidity
    function getSecret() public onlyOwner view returns(uint256 secret) {
        return 123;
    }
```

has the same output ABI as a struct with a single item in it.

It makes sense to just return the single value in this case, but the user loses that (pointless) name given to it

**Comment by fubuloubu** [comment]
> It makes sense to just return the single value in this case, but the user loses that (pointless) name given to it

Yes, I think it does make more sense to return the single value, however in this scenario would it still be a "tuple" at the top level? I'd think not

Also as an aside, a common way that named outputs is often used in Solidity code is like the following

```solidity
function getSecret() public onlyOwner view returns(uint256 secret) {
        ...
        secret = 123; // Output is now set
        ...
    }
```

File: tests/functional/conftest.py [hunk 4]
```diff
@@ -180,15 +401,46 @@ def owner(test_accounts):
...
+
+
+@pytest.fixture(params=("solidity", "vyper"))
+def contract_container(
+    request, solidity_contract_container, vyper_contract_container, networks_connected_to_tester
+):
+    return solidity_contract_container if request.param == "solidity" else vyper_contract_container
+
+
+@pytest.fixture(params=("solidity", "vyper"))
```
Comment (unparalleled-js) [inline tests/functional/conftest.py:444]: All tests that use this fixture use both a contract compiled with solidity as well as one with vyper and they have almost exactly the same ABI so they work for most tests.

**Review by NotPeopling2day — APPROVED** [review]
LGTM

**Comment by unparalleled-js** [comment]
I just noticed this PR does not handle struct output.
I am wondering if `EcosystemAPI` needs change for that to work properly.. I see there is a TODO in `contracts/base.py` for handling struct output

File: src/ape_ethereum/ecosystem.py [hunk 5]
```diff
@@ -88,6 +91,34 @@ class Block(BlockAPI):
     """
 
 
+def parse_output_type(output_type: str) -> Union[str, Tuple]:
```
Comment (fubuloubu) [inline src/ape_ethereum/ecosystem.py:94]: If you're unsure about a recursive function, it's usually a good indication to introduce fuzzing into the mix. Perhaps something like:

```py
from hypothesis import strategies as st

PRIMITIVE_TYPES = ["uint256", "address", ...]

@given(complex_type=st.tuples(st.one_of(PRIMITIVE_TYPES), min_length=1, max_length=20))
def test_output_parsing(complex_type):
    assert parse_output_type(",".join(complex_type)) == complex_type
```

**Comment by unparalleled-js** [comment]
- [ ] Edge case that is not currently ideal: tuples with structs in them. The entire thing gets interpreted as a single tuple because of ABI strangeness .. Going to look into possible solutions

File: src/ape_ethereum/ecosystem.py (hunk 5 above)
Comment (unparalleled-js) [reply src/ape_ethereum/ecosystem.py:94]: I added a test; it helped me find an edge case.

File: src/ape_ethereum/structs.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
+from dataclasses import make_dataclass
+from typing import Any, List, Optional, Union
+
+from ethpm_types.abi import ABIType, MethodABI
+
+
+class StructParser:
```
Comment (fubuloubu) [inline src/ape_ethereum/structs.py:7]: Can we move this to `ape.utils`?

File: src/ape/utils/misc.py [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+    """
+    Injected properties are injected class variables that must be set before use
+    **NOTE**: do not appear in a Pydantic model's set of properties.
+    """
+
+    def __get__(self, *args):
+        raise ValueError("Value not set. Please inject this property before calling.")
+
+
+class ManagerAccessMixin:
```
Comment (fubuloubu) [inline src/ape/utils/misc.py:247]: Feel like this stuff and below should be in their own separate modules

**Comment by fubuloubu** [comment]
Looks good!

**Review by unparalleled-js — APPROVED** [review]
